app/assets/javascripts/govuk_publishing_components/analytics/auto-scroll-tracker.js

```diff
@@ -0,0 +1,184 @@
+  }
+
+  AutoScrollTracker.prototype.trackVisibleNodes = function () {
+    for (var i = 0; i < this.trackedNodes.length; i++) {
+      var node = this.trackedNodes[i]
+      if (this.isVisible(node.top, node.bottom) && !node.alreadySeen) {
+        node.alreadySeen = true
+        // we store whether a heading has been tracked or not on the heading
+        // because if headings appear/disapper (e.g. inside an accordion)
+        // the order changes, so we can't refer to the previous trackedNodes
+        // as we do with percentages
+        if (node.node) {
+          node.node.setAttribute('data-autoscrolltracker-already-seen', true)
+        }
+
+        var action = node.eventData.action
+        var label = node.eventData.label
+
+        if (window.GOVUK.analytics && window.GOVUK.analytics.trackEvent) {
+          window.GOVUK.analytics.trackEvent('ScrollTo', action, { label: label, nonInteraction: true })
+        }
+      }
+    }
+  }
+
+  AutoScrollTracker.prototype.isVisible = function (top, bottom) {
+    return window.scrollY <= top && (window.scrollY + this.windowHeight) >= bottom
+  }
+
+  AutoScrollTracker.Heading = function (config) {
+    this.config = config
+  }
+
+  AutoScrollTracker.Heading.prototype.getTrackingNodes = function () {
+    var headingsDetails = []
+    var headings = ['h1', 'h2', 'h3', 'h4', 'h5', 'h6']
+    var headingsFound = document.querySelectorAll(headings)
+    if (headingsFound) {
+      for (var i = 0; i < headingsFound.length; i++) {
+        var heading = headingsFound[i]
+        // only track headings that are visible i.e. not inside display: none
+        if (heading.offsetParent !== null && !this.shouldBeIgnored(heading)) {
+          var pos = heading.getBoundingClientRect()
+          headingsDetails.push({
+            node: heading,
+            alreadySeen: heading.getAttribute('data-autoscrolltracker-already-seen'),
+            top: pos.top + document.documentElement.scrollTop,
+            bottom: pos.bottom + document.documentElement.scrollTop,
+            eventData: { action: 'Heading', label: heading.textContent.replace(/\s+/g, ' ').trim() }
+          })
+        }
+      }
+    }
+    return headingsDetails
+  }
+
+  // check heading isn't inside something that needs ignoring e.g. cookie banner, footer
+  AutoScrollTracker.Heading.prototype.shouldBeIgnored = function (heading) {
+    for (var i = 0; i < this.config.ignoreHeadingsInside.length; i++) {
+      if (heading.closest(this.config.ignoreHeadingsInside[i])) {
+        return true
+      }
+    }
+  }
+
+  AutoScrollTracker.Percentage = function (config) {
+    this.config = config
+  }
+
+  AutoScrollTracker.Percentage.prototype.getTrackingNodes = function (trackedNodes) {
+    var body = document.body
+    var html = document.documentElement
+    var pageHeight = Math.max(body.scrollHeight, body.offsetHeight, html.clientHeight, html.scrollHeight, html.offsetHeight)
```

Comment:
Corr, that's a lot of potential heights!

Reply:
Yeah, apparently the only way to be sure...